mp4/mdat.go

```diff
@@ -84,3 +87,42 @@ func (m *MdatBox) HeaderSize() uint64 {
 func (m *MdatBox) PayloadAbsoluteOffset() uint64 {
 	return m.StartPos + m.HeaderSize()
 }
+
+// AddReadSeeker - add readseeker to Mdat box.
+// When a file is decoded lazily, the Mdat Data byte slice is nil
+// and this readseeker is to read data whenever the data is needed.
+func (m *MdatBox) AddReadSeeker(rs io.ReadSeeker) {
+	m.readSeeker = rs
+}
+
```

Comment:
I think this should be called "SetReadSeeker" since there cannot be more than one, and this should probably be called automatically in DecodeFile?

Reply:
thanks for the suggestion. I am very bad at naming things. I changed it to unexported method because I think it should not be set outside of the decode process. Also, it will be invoked in the DecodeFile based on the DecMode.

Reply:
I think it should be possible to set it later. I'm thinking about a case of an on-the-fly progressive to DASH OnDemand translator. Then the first request would read the moov box, cache its data structure and generate a manifest with byte ranges (via a sidx box). Subsequent requests may be done by other threads which will use the moov data structures to generated moof boxes, and finally fill in media data from the mdat. In that case, it would be good if each thread could have its own ReadSeeker so that multiple thread can work on the same asset. I saw that you introduced a mutex, I suppose that was introduced to solve the issue of multiple threads using the same mdat object. Maybe the best is to be able to pass the readseeker as a parameter.